c10/util/intrusive_ptr.h

```diff
@@ -285,9 +285,11 @@ class intrusive_ptr final {
   // This constructor will not increase the ref counter for you.
   // We use the tagged dispatch mechanism to explicitly mark this constructor
   // to not increase the refcount
+ public:
```

Comment:
We should talk about this. I think we need this constructor for `OptionalTensorRef` to work as intended but it must be private for a good reason.

Reply:
I'm +1 for this constructor, it is tagged (so you have to explicitly use it) and I am going to claim there is no chesterton fence here.

Reply:
This seems like it should work. Are the pastes out of date?

Reply:
No, they aren't out of date. I didn't realize the refcount increments/decrements were inlined by `-O2` and show up in the assembly as `lock xaddq %rdx, 8(%rax)`, but after playing around with my code in the compiler explorer some more, I realized that the refcount increment that is visible in the assembly is from copying the `OptionalTensorRef` instance when calling `externalFunctionOptionalTensorRef`:
```
extern void externalFunctionOptionalTensorRef(OptionalTensorRef ref);

/**
Requirement #1
Given a const Tensor& in scope, produce an OptionalTensor (or reference to optional tensor) without bumping refcount.
*/
void createOptionalTensorRefFromTensor(const Tensor& src) {
    OptionalTensorRef refToSrc(src);
    externalFunctionOptionalTensorRef(refToSrc);   // refToSrc is *copied* to the parameter named ref
}
```
There is no corresponding refcount decrement because `~OptionalTensorRef` calls `release_()`. Adding a copy constructor for `OptionalTensorRef` that constructs `this->ref_` using this constructor of `c10::intrusive_ptr` that does not increase reference count solves the problem. I'll update the PR tomorrow.

Reply:
I updated the pastes with the new assembly.